#  Background

A general workflow for processing 16S rRNA gene amplicon data in R.

## Set your working directory in R
```
setwd("~/Desktop/microbiome/")
```
## Installing Packages

```
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install(c("DESeq2","readxl","openxlsx","vegan","ggplot2","dplyr","tidyverse","phyloseq","gridExtra","grid","ggpubr","agricolae","microbiome","MicrobeR","microbiomeSeq","apeglm"))
```

### Load the following libraries of the packages
```{r setup, include=FALSE}
library(DESeq2)
library(readxl)
library(openxlsx)
library(vegan)
library(ggplot2)
library(dplyr)
library(tidyverse)
library(phyloseq)
library(gridExtra)
library(grid)
library(ggpubr)
library(agricolae)
library(microbiome)
library(MicrobeR)
library(pheatmap)
library("apeglm")
library(ranacapa)
library(devtools)
library(phylosmith)
library(coin)
library(MicrobiotaProcess)
```

#### Files to download in R
 A. ASV count table
 B. Taxonomy metadata 
 C. Sample Metatable 
 
 ## Reading data into R
### The format of microbiome data
```
Most pipelines for processing high-throughput amplicon data, such as mothur, QIIME, and dada2, result in a matrix of read counts or OTU/ASV count tables.One dimension of this matrix (i.e. the rows or the columns) consists of OTUs or ASVs.The other dimension consists of samples.
Different tools will expect/output different orientations of the matrix, but, in our case, columns are samples and rows are OTUs/ASVs.
Sometimes the OTU/ASV data and the abundance matrix are two separate tables. There is usually another table with sample information in rows.
This makes it easy to add lots of additional sample data columns that can be used to subset the data. Each sample and OTU/ASV will have an unique ID.
```

### Load all the ASV counttables and metatable, taxonomy table you downloaded and combine all into one big table in R

###### Load ASV Count table
```{r}
abund_table <-read.xlsx("~/Desktop/Microbiome_files.xlsx", sheet="ASV", rowNames = TRUE)
abund_table<- as.data.frame(t(abund_table))
abund_table1 <- cbind(SampleID = rownames(abund_table), abund_table)
rownames(abund_table1) <- NULL
abun_count <-gather(abund_table1, FeatureID, Value, ASV_1:ASV_4155)
```

###### Load Metatable
```{r}
meta_table <-read.xlsx("~/Desktop/Microbiome_files.xlsx", sheet="meta_table", rowNames = TRUE)
meta_table1 <- cbind(SampleID = rownames(meta_table ), meta_table)
rownames(meta_table1 ) <- NULL
meta_table1$Treatment <- as.character(meta_table1$Treatment)
meta_table1$Treatment <- factor(meta_table1$Treatment, levels=unique(meta_table1$Treatment))
meta_table1$Trial <- as.character(meta_table1$Trial)
meta_table1$Trial <- factor(meta_table1$Trial, levels=unique(meta_table1$Trial))

```
###### Load Taxonomy table
```{r}
taxa1<-read_excel("~/Desktop/Microbiome_files.xlsx", sheet = "metadata")
tax_table1 <-do.call(rbind, strsplit(as.character(taxa1$Taxon),";"))
colnames(tax_table1) <-c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species")
rownames(tax_table1)<-taxa1$FeatureID
taxa_A <- as.data.frame(tax_table1)
taxa_A1  <- cbind(FeatureID = rownames(taxa_A ), taxa_A)
rownames(taxa_A1) <- NULL
```

## Joinn all data(count talbe, taxnonomy and meta table)
```
M1count<-full_join(taxa_A1, abun_count)
combine_b <-full_join(M1count,meta_table1)
```
###### Basic Statistics
```{r}
newdata <-combine_b %>% group_by(SampleID,FeatureID) %>% summarise(Value=mean(Value))
newdata1 <-spread(newdata, SampleID, Value) ##spread the data into a wide format
x<-dim(newdata1)[2] # number of columns
newdata1[2:x] <- lapply(newdata1[2:x], function(x) as.numeric(as.character(x)))
seq_total<-apply(newdata1[2:x],2,sum) #number of sequences per sample
ASV_count<-colSums(newdata1[2:x]>0) # ASVs per sample
ASV_single<-colSums(newdata1[2:x]==1) # ASVs with only 1 sequence
ASV_double<-colSums(newdata1[2:x]==2) # ASVs that have only 2 sequences
ASV_true<-colSums(newdata1[2:x]>2) # Number of ASVs with >2 sequences
dataset_info<-data.frame(seq_total,ASV_count)# combine the sequence total and ASV counts

#add the rownames as a proper column
dataset_info <- cbind(SampleID = rownames(dataset_info), dataset_info)
dataset_info
rownames(dataset_info) <- NULL
#combine the meta table with dataset_info
data_meta <- meta_table1 [c(1,2,4)]
basic_stat <- full_join(data_meta, dataset_info)
basic_stat

#write.csv(basic_stat,"basic_stat.csv") #Write your results to files
```

#### Convert data to a phyloseq object
```{r}
Abund_table<- as.data.frame(t(abund_table)) ##Transpose abund_table to have ASVs as rows and Sample IDs as columns
A2 <-as.matrix(Abund_table)

# First we need to create a phyloseq object using our count table
count_tab_pyA <- otu_table(A2, taxa_are_rows=T)

# Read in the taxonomy 
taxa_A2 <-taxa_A1 %>% remove_rownames %>% column_to_rownames(var="FeatureID")
taxa_A2 <-as.matrix(taxa_A2)
tax_tab_phy <- tax_table(taxa_A2)

# Read in the meta table
meta_table2 <-meta_table1 %>% remove_rownames %>% column_to_rownames(var="SampleID")
sample_info_tab_phy <- sample_data(meta_table2)

## Convert to phyloseq
All_S4 <-phyloseq(count_tab_pyA, tax_tab_phy, sample_info_tab_phy)
All_S4
```

### Data Filtering
##### Step1. Evaluate Amplicon Sequence Variants (ASV) summary statistics
#####  Remove taxa with zero counts 

```{r}
summary(taxa_sums(All_S4))
phy <- prune_taxa(taxa_sums(All_S4) > 0, All_S4)
```

```{r}
# Create a new data frame of the sorted row sums, a column of sorted values from 1 to the total number of individuals/counts for each ASV and a categorical variable stating these are all ASVs.
readsumsdf <- data.frame(nreads = sort(taxa_sums(phy), decreasing = TRUE), 
                        sorted = 1:ntaxa(phy), type = "ASVs")
# Make a data frame with a column for the read counts of each sample for histogram production
sample_sum_df <- data.frame(sum = sample_sums(phy))
```
#### Step2. Factor reordering and renaming
##### The default sorting for ggplot2 in R is alphabetical. For example if you want to label UV and Non-UV, It will by default place Non-UV on the left and UV on the right. However, you can switch the positions of the labels on the plot.

```{r}
# Relabel Water Type
levels(sample_data(phy)$Type)
sample_data(phy)$Type <- factor(sample_data(phy)$Type, levels = c("UV","NUV","UV1"))
levels(sample_data(phy)$Type)

# Relabel Name
sample_data(phy)$Name <- factor(sample_data(phy)$Name, labels = c("T6UV","T6NUV","T6UV1","T7UV"))
levels(sample_data(phy)$Name)

# Reorder Trial labels
levels(sample_data(phy)$Trial)
sample_data(phy)$Trial <- factor(sample_data(phy)$Trial, levels = c("Trial6", "Trial7"))
levels(sample_data(phy)$Trial)
```

#### Step3. Make plots for number of reads
###### Generates a bar plot with  number of reads (y-axis) for each taxa. Sorted from most to least abundant

```{r}
p.reads = ggplot(readsumsdf, aes(x = sorted, y = nreads)) +
  geom_bar(stat = "identity") +
  ggtitle("ASV Assessment") +
  scale_y_log10() +
  facet_wrap(~type, scales = "free") +
  ylab("# of Sequences")
p.reads
```

#### Step 4: Detect and remove outlier samples
##### Format a data table to combine sample summary data with sample variable data
```{r}
ss <- sample_sums(phy)
sd <- as.data.frame(sample_data(phy))
ss.df <- merge(sd, data.frame("Reads" = ss), by ="row.names")
# Plot the data by the treatment variable
y = 1000 # Set a threshold for the minimum number of acceptable reads. 
x = "SampleName" # Set the x-axis variable you want to examine
label = "Row.names" # This is the label you want to overlay on the points
p.ss.boxplot <- ggplot(ss.df, aes_string(x, y = "Reads", color = "Treatment")) + 
  geom_boxplot(outlier.colour="NA", position = position_dodge(width = 0.8)) +
  geom_jitter(size = 2, alpha = 0.6) +
  scale_y_log10() +
  facet_wrap(~Name) +
  geom_hline(yintercept = y, lty = 2) +
  geom_text(aes_string(label = label), size = 3, nudge_y = 0.05, nudge_x = 0.05)
p.ss.boxplot

```
The above data assessment helps to
1) Get the number of sequences per taxa (left plot). 
2) The number of reads per sample. 
3) Both of these plots will help give an understanding of how your data are structured across taxa.

```

#### Step 5: Taxon cleaning
###### The following R chunk removes taxa not-typically part of a bacterial microbiome analysis.
```{r}
get_taxa_unique(phy, "Kingdom")
get_taxa_unique(phy, "Class")
phy # Check the number of taxa prior to removal
ntaxa(phy)
ps2 <- phy %>%
  subset_taxa(
    Kingdom == "Bacteria" &
    Family  != "mitochondria" &
    Class   != "Chloroplast" &
      Phylum != "Chloroplast"&
      Order   != "mitochondria" &
      Order   != "Chloroplast"
  )
ps2 # Confirm that the taxa were removed
get_taxa_unique(ps2, "Kingdom")
get_taxa_unique(ps2, "Class")
get_taxa_unique(ps2, "Phylum")
```

#### Step 6: Data Subsetting
###### All samples
```{r}
# Subset data by Trial6
ps2.T6 <- subset_samples(ps2, Trial == "Trial6")
any(taxa_sums(ps2.T6) == 0) # In this case it is TRUE, so remove the zero's
ps2.T6 <- prune_taxa(taxa_sums(ps2.T6) > 0, ps2.T6)
any(taxa_sums(ps2.T6) == 0) # It should now be false

# Subset data by Trial7
ps2.T7 <- subset_samples(ps2, Trial == "Trial7")
any(taxa_sums(ps2.T7) == 0) # In this case it is TRUE, so remove the zero's
ps2.T7 <- prune_taxa(taxa_sums(ps2.T7) > 0, ps2.T7)
any(taxa_sums(ps2.T7) == 0) # It should now be false
```

###### Taxa Prevalence estimation
```{r}
# Calculate ASVs  prevalence across the data set
prevdf <- apply(X = otu_table(ps2),MARGIN = ifelse(taxa_are_rows(ps2), yes = 1, no = 2),FUN = function(x){sum(x > 0)})
# Add taxonomy and total read counts to prevdf
prevdf <- data.frame(Prevalence = prevdf, TotalAbundance = taxa_sums(ps2), tax_table(ps2))
```
###### Taxa Prevalence plot
```{r}
prevdf1 <- subset(prevdf, Phylum %in% get_taxa_unique(ps2, "Phylum"))
p.prevdf1 <- ggplot(prevdf1, aes(TotalAbundance, Prevalence / nsamples(ps2),color=Family)) +
geom_hline(yintercept = 0.05, alpha = 0.5, linetype = 2) +
geom_point(size = 3, alpha = 0.7) +
scale_x_log10() +
xlab("Total Abundance") + ylab("Prevalence") +
facet_wrap(~Phylum) +
theme(legend.position="none") +
 ggtitle("Phylum Prevalence in All Samples\nColored by Family")
p.prevdf1

```
#### Data transformations methods and plots
Many analysis in community ecology benefit from data transformation. Many microbiome data sets do not fit to a normal distribution, but transforming them towards normality may enable more appropriate data for specific statistical tests.

###### Transform to Relative abundances

```{r}
ps2.ra <- transform_sample_counts(ps2, function(OTU) OTU/sum(OTU))
```
###### Log transformation 
```{r}
ps2.log <- transform_sample_counts(ps2, function(x) log(1 + x))
```

###### Histograms of the non-transformed data vs. the transformed data can address the shift to normality
```{r}
p.no.log <- qplot(rowSums(otu_table(ps2))) + ggtitle("Raw Counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of ASV")
p.log <- qplot(log10(rowSums(otu_table(ps2)))) +
  ggtitle("log10 transformed counts") +
  theme_bw() +
  xlab("Row Sum") +
  ylab("# of ASV")
ggarrange(p.no.log, p.log, ncol = 2, labels = c("A)", "B)"))
```
#### Rarefaction plot
##### We can plot the rarefaction curve for the observed ASVs in the entire data set. This is a way to check the richness captured in the sequencing effort.
```{r}
Raf <-ggrare(ps2, step = 24, color="Name",  se = TRUE)
```
### Alpha diversity

######Alpha diversity is a standard tool researchers can use to calculate the number of bacterial taxa present in a study or study group and the relationships between relative abundance and how evenly taxa are distributed. These are classic representations of species number and diversity in a study which provide useful summary information about the numbers and relative abundances of bacterial taxa within the study.

#We can calculate several measures of alpha diversity, add them to a data frame and use ggplot2 to observe the different alpha diversity metrix.
```{r}
SD <- sample_data(ps2) %>%
    data.frame() %>%
    select("Treatment","Trial","Type", "Name") %>%
    mutate_if(is.factor,as.character)
SD1 <- cbind(SampleID = rownames(SD ), SD)
rownames(SD1 ) <- NULL
All_S4_prune_rarefy <- rarefy_even_depth(ps2, rngseed= 81, sample.size = 10000)
adiv <- data.frame(
  "Observed" = phyloseq::estimate_richness(All_S4_prune_rarefy, measures = "Observed"),
  "Shannon" = phyloseq::estimate_richness(All_S4_prune_rarefy, measures = "Shannon"),
  "Simpson" = phyloseq::estimate_richness(All_S4_prune_rarefy, measures = "Simpson"),
  "Chao1"= phyloseq::estimate_richness(All_S4_prune_rarefy, measures = "Chao1"),
  "Status" = phyloseq::sample_data(All_S4_prune_rarefy)$Treatment)
head(adiv)
adiv$Trial <- SD$Trial
adiv$Treatment <- SD$Treatment
adiv$Name <- SD$Name
adiv

```
##### Simpson diversity
```{r}
Div_SIM <- ggplot(adiv, aes(as.factor(Name), Simpson)) +
  facet_grid(.~Trial, scales="free",space="free_x",switch="y")+
  geom_boxplot(aes(fill=factor(Treatment)),width=0.5)+
  scale_fill_manual(values=c("blue", "orange","blue", "orange")) +
  labs(y="Simpson's Diversity", x="Trial", fill="Treatment")+theme_bw()+
  theme(strip.text.x = element_text(size=12,face="bold"))+
  theme(strip.background = element_rect(fill="gray85"))+
  theme(axis.text.y = element_text(size="14", color="black"), axis.title.y = element_text(face="bold",size="14", color="black"))+theme(axis.text.x = element_text(size="12", color="black", angle = 45), axis.title.x.bottom = element_text(face="bold",size="14", color="black"))+theme(axis.text.x.top= element_text(face="bold",size="14", color="black"),
        legend.position = "bottom",panel.spacing = unit(0.2, "mm"),               
        panel.border = element_rect(colour="grey"))+
  scale_y_continuous(limits=c(0,1))
Div_SIM

## Statistics
Simp.aov <- aov(Simpson ~ Name:Treatment+Name+Treatment, data = adiv)
summary(Simp.aov)
TukeyHSD(Simp.aov)
tukey_Trial2 <-HSD.test(Simp.aov, "Trial", group = TRUE)
tukey_Trial2
kruskal.test(Simpson ~ Name, data=adiv)
```
##### Richness
```{r}
Div_rich <- ggplot(adiv, aes(as.factor(Name), Chao1.Chao1)) +
  facet_grid(.~Trial, scales="free",space="free_x",switch="y")+
  geom_boxplot(aes(fill=factor(Treatment)),width=0.5)+
  scale_fill_manual(values=c("blue", "orange","blue", "orange")) +
  labs(y="Chao1 Richness", x="Trial", fill="Treatment")+theme_bw()+
  theme(strip.text.x = element_text(size=12,face="bold"))+
  theme(strip.background = element_rect(fill="gray85"))+
  theme(axis.text.y = element_text(size="14", color="black"), axis.title.y = element_text(face="bold",size="14", color="black"))+theme(axis.text.x = element_text(size="14", color="black", angle = 45), axis.title.x.bottom = element_text(face="bold",size="14", color="black"))+theme(axis.text.x.top= element_text(face="bold",size="14", color="black"),
        legend.position = "bottom",panel.spacing = unit(0.2, "mm"),               
        panel.border = element_rect(colour="grey"))+
  scale_y_continuous(limits=c(0,300))
Div_rich
##Statistics
Rich.aov <- aov(Chao1.Chao1 ~ Name:Treatment+Name, data = adiv)
summary(Rich.aov)
tukey_Trial2 <-HSD.test(Rich.aov, "Name", group = TRUE)
TukeyHSD(Rich.aov)
tukey_Trial2
kruskal.test(Chao1.Chao1 ~ Treatment, data=adiv)
kruskal.test(Chao1.Chao1 ~ Name, data=adiv)
```

#### NMDS Plot
#### Beta diversity enables you to view overall relationships between samples. These relationships are calculated using a distance metric calculation (of which there are many) and these multi-dimensional relationships are evaluated and viewed in the two dimensions which explain the majority of their variance. Additional dimensions can be explored to evaluate how samples are related to one another.

###### Log transformation moves to a more normal distribution
```{r}
GPA1 <- ps2 %>% transform_sample_counts(function(x) log(1 + x))%>% psmelt()
all_order <- GPA1 %>%group_by(Sample,OTU)%>%summarise(avg_abundance = mean(Abundance))
Oder <-spread(all_order, Sample,avg_abundance) ##spread the data into a wide format
Order <-Oder %>% remove_rownames %>% column_to_rownames(var="OTU")
Abun <- as.data.frame(t(Order)) 
SD 

```

```{r}
sol_t1 <-metaMDS(Abun ,distance = "bray", k = 2, trymax = 50)
NMDS_t1=data.frame(NMDS1=sol_t1$point[,1],NMDS2=sol_t1$point[,2],
          Treatment=as.factor(SD[,1]),Name=as.factor(SD[,4]),Type=as.factor(SD[,3]))

veganCovEllipse<-function (cov, center = c(0, 0), scale = 1, npoints = 100) 
{
  theta <- (0:npoints) * 2 * pi/npoints
  Circle <- cbind(cos(theta), sin(theta))
  t(center + scale * t(Circle %*% chol(cov)))
}
# define shapes for plots
shape_values<-seq(1,11)
# set theme for following plots
theme_set(theme_gray())
plot.new()
ord_t1<-ordiellipse(sol_t1, as.factor(SD$Type),
                    display = "sites", kind ="sd", conf = 0.95, label = T)
dev.off()
#Generate ellipse points based on 95% confidence (SD) intervals
df_ell_t1 <- data.frame()
for(g in levels(NMDS_t1$Type)){
  if(g!="" && (g %in% names(ord_t1))){
    df_ell_t1 <- rbind(df_ell_t1,
    cbind(as.data.frame(with(NMDS_t1[NMDS_t1$Type==g,], veganCovEllipse(ord_t1[[g]]$cov,ord_t1[[g]]$center,ord_t1[[g]]$scale))),Type=g))}}
#Store the center of the ellipses (mean NMDS)
NMDS.mean_t1=aggregate(NMDS_t1[,1:2],list(group=NMDS_t1$Type),mean)
#Calculate p-value:
adon_t1<-adonis2(Abun ~Name, data=SD, by=NULL,method="bray", k=2)
#Plot NMDS
NMDS_t1 <- cbind(SampleID = rownames(NMDS_t1 ), NMDS_t1)
rownames(SD1 ) <- NULL
NMDSplot_t3<-ggplot(data=NMDS_t1,aes(NMDS1,NMDS2,col=Type))+
  #geom_text(data=NMDS_t1,aes(NMDS1,NMDS2,label=SampleID),size=3,vjust=0)+
  # label the middle of the ellipses with the name of the grouping factor
 # annotate("text",x=NMDS.mean_t1$x,y=NMDS.mean_t1$y,
          # label=NMDS.mean_t1$group,size=5)+
  # add the p-value in the bottom right corner
  annotate("text",x=min(NMDS_t1$NMDS1),y=min(NMDS_t1$NMDS2-0.5),
           label=paste("p= ", adon_t1$`Pr(>F)`[1]),size=3)+
  # draw the ellipses. define color based on the grouping factor
  geom_path(data=df_ell_t1, aes(x=NMDS1, y=NMDS2, linetype=Type), size=1)+
  scale_linetype_manual(values=c("UV"="solid", "NUV"="longdash","UV1"="dotdash"))+
  scale_colour_manual(values=c("UV"="gold2", "NUV"="hotpink4","UV1"="blue"))+
  ggtitle("By Type")+
  # add the points per sample. define shape based on Day
  geom_point(aes(shape=Treatment), size=5) + scale_shape_manual(values=c(15,16)) +
  # reorder the legend
  guides(color = guide_legend(order=1),lty= guide_legend(order=1),
         shape = guide_legend(order=2), legend.position = "NONE")
NMDSplot_Hatchery <- NMDSplot_t3 + theme(axis.text.y = element_text(size="12", color="black"), axis.title.y = element_text(face="bold",size="12", color="black"))+
theme(axis.text.x = element_text(size="12", color="black"), axis.title.x.bottom = element_text(face="bold",size="12", color="black"))+
theme(axis.text.x.top= element_text(face="bold",size="12", color="black"))+theme(plot.title=element_text(size = 16, face = "bold", hjust = 0.5))
NMDSplot_Hatchery
```

#### Community composition plotting
Classic bar plots of bacterial composition present in each sample can be useful for communicating results. These are relatively easy to interpret when major shifts in microbial communities are present, however, they are not effective at detecting subtle shifts in communities or taxa and do not convey any statistical significance. It is appropriate to subject these plots to further statistical analysis. 

#### Calculate number of reads and add it to the meta table
```
sample_sum_df <- data.frame(sum = sample_sums(ps2 ))
sum(sample_sum_df[, 'sum'])

Sample_reads <- cbind(SampleID = rownames(sample_sum_df), sample_sum_df)
rownames(Sample_reads) <- NULL

#Join all the data
Meta_T <-full_join(meta_table1, Sample_reads)
```

##### Plot by Phylum 

```{r}
SD <- sample_data(ps2) %>%
    data.frame() %>%
    select("Treatment","Trial","Type", "Name") %>%
    mutate_if(is.factor,as.character)
SD1 <- cbind(SampleID = rownames(SD ), SD)
rownames(SD1 ) <- NULL
Orderabundance <- ps2 %>%
tax_glom(taxrank = "Phylum") %>% # agglomerate at phylum level
transform_sample_counts(function(x) {x/sum(x)*100}) %>% # Transform to rel. abundance
  psmelt() %>% 
  #filter(Abundance > 0.01)%>% # Melt to long format
  arrange(Phylum)
head(Orderabundance)
all_order <- Orderabundance %>%group_by(Sample,Phylum)%>%summarise(avg_abundance = mean(Abundance))
Oder <-spread(all_order, Sample,avg_abundance) ##spread the data into a wide format
Order <-Oder %>% remove_rownames %>% column_to_rownames(var="Phylum")
Order1c<- as.data.frame(t(Order))
Order1b <-Order1c[,order(colSums(Order1c),decreasing=TRUE)]
#Extract list of top 12 Taxa and assign the remaining as "Others"
N<-4 # can change number of taxa to plot here
taxa_list<-colnames(Order1b)[1:N]
#Generate a new table with everything added to Others
Order1d <-data.frame(Order1b[,colnames(Order1b) %in% taxa_list], Others=rowSums(Order1b[,!colnames(Order1b) %in% taxa_list]))
Order1d_long <-gather(Order1d, SampleID, Value)
Order1d_long$Taxa <-factor(Order1d_long$SampleID, levels = unique(Order1d_long$SampleID))
#Make a new data fram called meta_taxa and merge the metadata and long taxa data
Order1d_count<-as.data.frame(c(SD1, Order1d_long))
###Get the colours to use for plot
my_colors <- c("#4DAA57","#F1A66A","#F26157", "#F9ECCC", "#679289", "#33658A","#F6AE2D","#86BBD8")
Order1d_count$Name <- as.character(Order1d_count$Name)
Order1d_count$Name <- factor(Order1d_count$Name, levels=c("T6UV","T6NUV","T6UV1","T7UV"))
Order1d_count$Treatment <- as.character(Order1d_count$Treatment)
Order1d_count$Treatment <- factor(Order1d_count$Treatment, levels=unique(Order1d_count$Treatment))
Order1d_count$Type <- as.character(Order1d_count$Type)
# plot the percent ASV abundances
Order_plot <- ggplot(Order1d_count)+
  geom_col(mapping = aes(x = SampleID, y = Value, fill = Taxa), position = "fill", show.legend = TRUE)+
  ylab("Percent Phylum Abundance") +
  theme(strip.background = element_rect(colour="black", size=1))+
  theme(strip.text.x = element_text(size=30,face="bold"))+
  facet_grid(.~Name+Treatment+Type, drop=TRUE,scale="free", space="free",switch="x")+
  scale_fill_manual(values = my_colors) +
  xlab("ProjectName")+
  scale_y_continuous(expand = c(0,0),labels=scales::percent)+
  theme_minimal()+
   theme(axis.text.x = element_blank()) +
  theme(strip.background = element_rect(fill="gray85"),panel.spacing = unit(0.2,"lines"),axis.text.y.left = element_text(colour="black", size=14, face="bold"),
        axis.title.y = element_text(size = 16, face="bold"),
        title = element_text(size = 16, face="bold"))+ theme(legend.position="none")+ theme(legend.title = element_text(colour="black", size=10))+ theme(legend.text = element_text(colour="black", size=14, face="bold"))
  labs(x="ProjectName", y="Percentage Phylum Abundance")

# plot the number of reads per sample
Read_plot <-ggplot(Meta_T,aes(SampleID,sum, fill=Treatment))+
  # plot using columns
  geom_col()+theme_grey()+
  # facet by type, then trial, then day, then treatment group.
  facet_grid(.~Name+Treatment+Type, drop=TRUE, scale="free_x")+
  # label the axes
  labs(y="Read Abundance", x=NULL)+
  # change the fill color to Control= light blue, Treatment=dark red.
  scale_fill_manual(values=c("gray67","gray5","#640025","#3ddda5","#cdffaa","#150e00","#bae278"))+
  scale_y_continuous(expand = c(0,0),labels=scales::scientific_format(digits=1))+
  # edit the facets a little
  theme(strip.text = element_blank(),legend.position="none",
        axis.ticks.x = element_blank(),
        panel.spacing = unit(0.4,"lines"),axis.text.x = element_blank())+
  theme(axis.text.y = element_text(size="14", color="black",face = "bold"), axis.title.y = element_text(face="bold",size="16", color="black"))

# get the legend for each plot
g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}
ASV_legend<-g_legend(ggplot(Order1d_count,aes(SampleID,Value,fill=Taxa))+ geom_col()+scale_fill_manual(values= my_colors)+labs(fill="Taxa"))
read_legend<-g_legend(ggplot(Meta_T,aes(SampleID,sum, fill=Treatment))+
                     geom_col()+scale_fill_manual(values=c("gray67","gray5","#640025","#3ddda5","#cdffaa","#150e00","#bae278"))+labs(fill="Treatment"))

# plot the figures and legends together
# plot the figures and legends together
Order_plot1 <-grid.arrange(Read_plot, read_legend, Order_plot, ASV_legend, nrow=2, ncol=2, heights=c(1,2.5), widths=c(9,3.5))
```

##### Plot by Order
```{r}
SD <- sample_data(ps2) %>%
    data.frame() %>%
    select("Treatment","Trial","Type", "Name") %>%
    mutate_if(is.factor,as.character)
SD1 <- cbind(SampleID = rownames(SD ), SD)
rownames(SD1 ) <- NULL
Orderabundance <- ps2 %>%
tax_glom(taxrank = "Order") %>% # agglomerate at phylum level
transform_sample_counts(function(x) {x/sum(x)*100}) %>% # Transform to rel. abundance
  psmelt() %>% 
  #filter(Abundance > 0.01)%>% # Melt to long format
  arrange(Order)
head(Orderabundance)
all_order <- Orderabundance %>%group_by(Sample,Order)%>%summarise(avg_abundance = mean(Abundance))
Oder <-spread(all_order, Sample,avg_abundance) ##spread the data into a wide format
Order <-Oder %>% remove_rownames %>% column_to_rownames(var="Order")
Order1c<- as.data.frame(t(Order))
Order1b <-Order1c[,order(colSums(Order1c),decreasing=TRUE)]
#Extract list of top 12 Taxa and assign the remaining as "Others"
N<-12 # can change number of taxa to plot here
taxa_list<-colnames(Order1b)[1:N]
#Generate a new table with everything added to Others
Order1d <-data.frame(Order1b[,colnames(Order1b) %in% taxa_list], Others=rowSums(Order1b[,!colnames(Order1b) %in% taxa_list]))
Order1d_long <-gather(Order1d, SampleID, Value)
Order1d_long$Taxa <-factor(Order1d_long$SampleID, levels = unique(Order1d_long$SampleID))
#Make a new dataframe called meta_taxa and merge the metadata and long taxa data
Order1d_count<-as.data.frame(c(SD1, Order1d_long))
###Get the colours to use for plot
my_colors <- c(
  "#CBD588", "#5F7FC7", "orange","#DA5724", "#508578", "#CD9BCD",
  "#AD6F3B", "#673770","#D14285", "#652926", "#C84248",
  "#8569D5", "#5E738F","#D1A33D", "#8A7C64", "#599861", "#808080","#94FFB5","#A54657",  "#582630", "#F7EE7F", "#4DAA57","#F1A66A","#F26157", "#F9ECCC", "#679289", "#33658A",
"#F6AE2D","#86BBD8")
Order1d_count$Name <- as.character(Order1d_count$Name)
Order1d_count$Name <- factor(Order1d_count$Name, levels=c("T6UV","T6NUV","T6UV1","T7UV"))
Order1d_count$Treatment <- as.character(Order1d_count$Treatment)
Order1d_count$Treatment <- factor(Order1d_count$Treatment, levels=unique(Order1d_count$Treatment))
Order1d_count$Type <- as.character(Order1d_count$Type)
# plot the percent ASV abundances
Order_plot <- ggplot(Order1d_count)+
  geom_col(mapping = aes(x = SampleID, y = Value, fill = Taxa), position = "fill", show.legend = TRUE)+
  ylab("Percent Order Abundance") +
  # facet_wrap(. ~Trial+Type, scales = "free_x") +
  theme(strip.background = element_rect(colour="black", size=1))+
  theme(strip.text.x = element_text(size=20,face="bold"))+
  facet_grid(.~Name+Treatment+Type, drop=TRUE,scale="free", space="free",switch="x")+
  scale_fill_manual(values = my_colors) +
  xlab("Trial")+
  scale_y_continuous(expand = c(0,0),labels=scales::percent)+
  theme_minimal()+
   theme(axis.text.x = element_blank()) +
  theme(strip.background = element_rect(fill="gray85"),panel.spacing = unit(0.2,"lines"),axis.text.y.left = element_text(colour="black", size=14, face="bold"),
        axis.title.y = element_text(size = 14, face="bold"),
        title = element_text(size = 14, face="bold"))+ theme(legend.position="bottom")+ theme(legend.title = element_text(colour="black", size=10))+ theme(legend.text = element_text(colour="black", size=14, face="bold"))
  labs(x="Trial", y="Percentage Phylum Abundance")

# plot the number of reads per sample
Read_plot <-ggplot(Meta_T,aes(SampleID,sum, fill=Treatment))+
  # plot using columns
  geom_col()+theme_grey()+
  # facet by type, then trial, then day, then treatment group.
  facet_grid(.~Name+Treatment+Type, drop=TRUE, scale="free_x")+
  # label the axes
  labs(y="Read Abundance", x=NULL)+
  # change the fill color to Control= light blue, Treatment=dark red.
  scale_fill_manual(values=c("gray67","gray5","#640025","#3ddda5","#cdffaa","#150e00","#bae278"))+
  scale_y_continuous(expand = c(0,0),labels=scales::scientific_format(digits=1))+
  # edit the facets a little
  theme(strip.text = element_blank(),legend.position="none",
        axis.ticks.x = element_blank(),
        panel.spacing = unit(0.4,"lines"),axis.text.x = element_blank())+
  theme(axis.text.y = element_text(size="14", color="black",face = "bold"), axis.title.y = element_text(face="bold",size="14", color="black"))

# get the legend for each plot
g_legend<-function(a.gplot){
  tmp <- ggplot_gtable(ggplot_build(a.gplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)}
ASV_legend<-g_legend(ggplot(Order1d_count,aes(SampleID,Value,fill=Taxa))+  geom_col()+scale_fill_manual(values= my_colors)+labs(fill="Taxa"))
read_legend<-g_legend(ggplot(Meta_T,aes(SampleID,sum, fill=Treatment))+
                     geom_col()+scale_fill_manual(values=c("gray67","gray5","#640025","#3ddda5","#cdffaa","#150e00","#bae278"))+labs(fill="Treatment"))
# plot the figures and legends together
Order_plot1 <-grid.arrange(Read_plot, read_legend, Order_plot, ASV_legend, nrow=2, ncol=2, heights=c(1,2.5), widths=c(9,3.5))
```

# Differential Abundance
There are a number of ways to detect bacterial taxa (more accurately, ASVs) differentially abundant between one or more groups or along an environmental gradient (e.g. random forest, LEfSe, ANCOM, DESEQ2)

#### Plot using LEFSe
```
xt <- transform(ps2, 'hellinger')
#Orderabundance <- xt %>% tax_glom(taxrank = "Order")
deres <- diff_analysis(obj =xt, classgroup = "Trial",
                       mlfun = "lda",alltax = FALSE,
                       filtermod = "pvalue",
                       firstcomfun = "kruskal_test",
                       firstalpha = 0.01,
                       strictmod = TRUE,
                       secondcomfun = "wilcox_test",
                       subclmin = 4,
                       subclwilc = TRUE,
                       secondalpha = 0.01,
                       lda=2)
es_pb <- ggeffectsize(obj=deres, 
                     lineheight=0.3,
                     linewidth=0.2,   pointsize = 4.5,) + 
        scale_color_manual(values=c("blue","orange",
                                    "orange","red", "chocolate4","gray0", "cyan", "green", "pink1","#599861"))+theme(axis.text.x = element_text(colour="black", size=20, face="bold")) +
  theme(strip.background = element_rect(fill="gray85"),panel.spacing = unit(0.2,"lines"),axis.text.y.left = element_text(colour="black", size=20, face="bold"),axis.text.y.right = element_text(colour="black", size=20, face="bold"),
        axis.title.y = element_text(size = 20, face="bold"),axis.title.x = element_text(size = 20, face="bold", colour = "black"),
        title = element_text(size = 20, face="bold"))+labs(title = "Trial")+ylab("Taxa")+theme(legend.position="bottom")+ theme(legend.title = element_text(colour="black", size=10))+ theme(legend.text = element_text(colour="black", size=14, face="bold"))
es_pb
```

